oak_loader/src/attestation.rs

```diff
@@ -0,0 +1,104 @@
+//
+// Copyright 2021 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use anyhow::Context;
+use http::uri::Uri;
+use log::info;
+use oak_proxy_attestation::proto::{
+    proxy_attestation_client::ProxyAttestationClient, GetSignedCertificateRequest,
+};
+use oak_sign::get_sha256_hex;
+use openssl::{
+    hash::MessageDigest,
+    pkey::{PKey, Private},
+    rsa::Rsa,
+    x509::{X509NameBuilder, X509ReqBuilder},
+};
+use tonic::{
+    transport::{Certificate, Channel, ClientTlsConfig, Identity},
+    Request,
+};
+
+/// Placeholder function for collecting measurements of TEEs, i.e. hashes of enclaves or VMs.
+fn get_tee_measurements() -> String {
+    get_sha256_hex("TEE".as_bytes())
+}
+
+/// .
+pub async fn get_tls_identity_from_proxy(
+    uri: &Uri,
+    root_tls_certificate: &[u8],
+) -> anyhow::Result<Identity> {
+    // Create unsigned certificate.
+    let rsa_key_pair = Rsa::generate(2048)?;
+    let private_key = rsa_key_pair.private_key_to_pem()?.to_vec();
+    let key_pair = PKey::from_rsa(rsa_key_pair)?;
+    let certificate_request = create_certificate_request(&key_pair)?;
+
+    // Request a signed certificate from the Proxy Attestation Service.
+    info!(
+        "Sending certificate signing request to Proxy Attestation Service: {:?}",
+        uri
+    );
+    let mut client = create_proxy_client(&uri, &root_tls_certificate)
+        .await
+        .context("Couldn't create gRPC client")?;
+    let request = Request::new(GetSignedCertificateRequest {
+        certificate_request,
+    });
+    let response = client
+        .get_signed_certificate(request)
+        .await
+        .context("Couldn't submit set")?;
+    let certificate = response.get_ref().certificate.to_vec();
+    info!("Received signed certificate from Proxy Attestation Service");
+
+    Ok(Identity::from_pem(&certificate, &private_key))
+}
+
+/// .
+fn create_certificate_request(key_pair: &PKey<Private>) -> anyhow::Result<Vec<u8>> {
+    let mut builder = X509ReqBuilder::new()?;
+    builder.set_pubkey(&key_pair)?;
+
+    let mut cert_name = X509NameBuilder::new()?;
+    // CN refers to Common Name.
+    // https://tools.ietf.org/html/rfc5280
+    cert_name.append_entry_by_text("CN", &get_tee_measurements())?;
```

Comment:
What makes sure that this is putting a real measurement in this field? AFAICT this is completely unrelated to the actual attestation with the proxy, so why would anyone trust it?

Reply:
The proxy will attest the backend and only sign a certificate that has a correct measurement in it.
But proxy probably should be the one who puts these measurements in the cert (in the end it will be identical cert).

Reply:
Is it not possible for the CA enclave to add those fields and sign them in its part of the cert, instead of double checking them from the self-signed cert from the worker enclave?